accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
+            };
+            ```
 
-    Subspace is simple to calculate. We already know that `L`'s type is a
-    subtype of `R` from earlier checks. `L` is a subspace if subtracting each
-    field in `R` from its corresponding field in `L` yields empty. For example:
+    *Note that in the common case where the sealed type hierarchy is not
+    generic, all of this simplifies to just being a list of spaces, one for each
+    type that is a direct subtype of the sealed type.*
 
-    ```
-    Jack(suit: heart, oneEyed: true) - Face(suit: Suit, oneEyed: bool)
+*   **`T` is an enum type:**
+
+    1.  For each element in the enum:
+
+        1.  If the element's type is a subtype of the overapproximation of the
+            enum type:
+
+            1.  A space with the type of the element and a constant restriction
+                with the element value.
+
+    *We have to check the type because with a generic enum, some elements might
+    not be subtypes of the matched type. Consider:*
+
+    ```dart
+    enum E<T> {
+      a<int>(),
+      b<String>(),
+      c<double>(),
+    }
+
+    method(E<num> e) {
+      switch (e) {
+        case E.a:
+        case E.c:
+          print('ok');
+      }
+    }
     ```
 
-    Here, `heart - Suit` is empty, as is `true - bool`, so `L` is a subspace.
+    *This switch is exhaustive because it's not possible for `E.b` to reach it.*
 
-5.  Else, subtract the field sets, below.
+    *We use overapproximation because the enum may refer to type parameters in
+    the surrounding code:*
 
-### Field set subtraction
+    ```dart
+    method<T extends num>(E<T> e) {
+      switch (e) {
+        case E.a:
+        case E.c:
+          print('ok');
+      }
+    }
+    ```
 
-We're finally at a point where have two objects `L` and `R` whose types allow
-them to be subtracted in such a way that their fields come into play. We also
-know there's no early out where the result is just `L` or empty.
+    *This is also exhaustive. When `method()` is instantiated with `int`, then
+    the second case will never match. Conversely, the first case will never
+    match in a call to `method<double>()`. But exhaustiveness must be sound
+    across all instantiations, so it uses overapproximation.*
 
-1.  Subtract each field in `R` from its corresponding field in `L`. We define
-    `fixed` to be the set of fields where that subtraction didn't change the
-    field and `changed` to be the fields where it did. To calculate these sets,
-    for each field in `L` or `R`:
+*   **`T` is nullable type `S?`:** A space of type `S` and a space of type
+    `Null`.
 
-    1.  Let `Lf` be the field in `L` and `Rf` be the field in `R` (either of
-        which may be inferred if not present).
+*   **`T` is `FutureOr<F>` for some `F`**: A space of type `Future<F>` and a
+    space of type `F`.
 
-    2.  Calculate `D = Lf - Rf`.
+*   **`T` is `bool`:** Two spaces of type `bool` with constant restrictions
+    `true` and `false`.
```

Comment:
And

*    **`T` is `Null`:** One space of type `bool` with the constant restriction `null`.

Reply:
I don't think that's necessary, but I could be wrong. The proposal tries to consistently use a space of type `Null` with no constant restriction to deal with `null`.